source/common/network/matching/inputs.cc

```diff
@@ -10,9 +11,10 @@ namespace Envoy {
 namespace Network {
 namespace Matching {
 
-template <>
-Matcher::DataInputGetResult DestinationIPInput<MatchingData>::get(const MatchingData& data) const {
-  const auto& address = data.socket().connectionInfoProvider().localAddress();
+template <class MatchingDataType>
+Matcher::DataInputGetResult
+DestinationIPInput<MatchingDataType>::get(const MatchingDataType& data) const {
```

Comment:
I think these template implementations should be inlined and moved to "inputs.h". Otherwise, we may face with linking issues.

Reply:
Can you elaborate? I've seen primary templates in `.cc` files used, and that works as long as it's used only within the `.cc` file.

Reply:
The original contains a full template specialization implementation for `MatchingData` and another for `UdpMatchingData`. A template specialization implementation can be located in `.cc`, and the linker will recognize their symbols for linking. Now, we are going to change to a template member implementation, the implementation should be moved to ".h". We may come up with

```cpp
// inputs.h
template <class MatchingDataType>
class SourceIPInput : public Matcher::DataInput<MatchingDataType> {
  // template member implementation
  Matcher::DataInputGetResult get(const MatchingDataType& data) const override {}
}

// template specialization definition for UdpMatchingData
template <>
Matcher::DataInputGetResult
DestinationIPInput<UdpMatchingData>::get(const UdpMatchingData& data) const;

// inputs.cc
// template specialization implementation for UdpMatchingData
template <>
Matcher::DataInputGetResult
DestinationIPInput<UdpMatchingData>::get(const UdpMatchingData& data) const {}
```